### ISSUE TITLE
Rule http_redirect_irule_443 or datagroup https_redirect_dg do not respect virtual-server.f5.com/http-port

### ISSUE THREAD

**Issue by bogiva** [F5Networks/k8s-bigip-ctlr#802]
#### Description

Neither the code at https://github.com/F5Networks/k8s-bigip-ctlr/blob/ae907570e02b6149ba6a636685fa1b9c53ea2eb9/pkg/appmanager/routing.go#L353 nor the code at https://github.com/F5Networks/k8s-bigip-ctlr/blob/55007560d27df820f1d5cea7ffd5b69599912d3a/pkg/appmanager/appManager.go#L977
do respect the value of the "virtual-server.f5.com/http-port" property.

If we set this property e.g. to 8080 then the redirect irule mentioned above does not work as expected, i.e. does not redirect to HTTPS.

#### Kubernetes Version
1.10.10

#### Controller Version
1.4.2

#### BIG-IP Version
12.1.2

**Comment by vincentmli** [comment]
according to https://devcentral.f5.com/wiki/iRules.getfield.ashx

[HTTP::host]  could be  <hostname>:<port>


maybe change 

<pre>                               
 if {$redir == 1} {
                                        HTTP::redirect https://[getfield [HTTP::host] ":" 1]:443[HTTP::uri]
                                }

</pre>
to

<pre>
                                if {$redir == 1} {
                                        HTTP::redirect https://[HTTP::host]:443[HTTP::uri]
                                }
</pre>

**Comment by vincentmli** [comment]
I did a test with following ingress using 'virtual-server.f5.com/http-port': '8080' and the redirect works fine, could you please let me know what I am missing here? any further detail from you?

<pre>
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: single-service
  annotations:
     ingress.kubernetes.io/ssl-redirect: "true"
     ingress.kubernetes.io/allow-http: "false"
     ingress.kubernetes.io/allow-https: "true"
     kubernetes.io/ingress.class: f5
     'virtual-server.f5.com/ip': '10.169.72.169'
     'virtual-server.f5.com/http-port': '8080'
spec:
  tls:
#  - secretName: test/test-clientssl
  - secretName: Common/clientssl
  backend:
    serviceName: nginxservice 
    servicePort: 80
</pre>

I have no problem with http_redirect_irule_443, my test with curl to visit 10.169.72.169:8080 and it got  redirect to https 443 virtual server as the alert from the certificate error

<pre>
# curl -L http://10.169.72.169:8080/
curl: (60) Issuer certificate is invalid.
................
If you'd like to turn off curl's verification of the certificate, use
 the -k (or --insecure) option.
</pre>

tcpdump to show the 302 redirection response
<pre>
# tcpdump -A -i 0.0 'host 10.169.72.169 and (((ip[2:2] - ((ip[0]&0xf)<<2)) - ((tcp[12]&0xf0)>>2)) != 0)'


13:05:44.867933 IP 10.169.72.93.57402 > 10.169.72.169.webcache: Flags [P.], seq 3306044799:3306044881, ack 3474184863, win 29200, options [nop,nop,TS val 595756839 ecr 361318307], length 82: HTTP: GET / HTTP/1.1 in slot1/tmm2 lis=/test/ingress_10-169-72-169_8080
E....[@.@...
.H]
.H..:....=.......r..N.....
#..'..G.GET / HTTP/1.1
User-Agent: curl/7.29.0
Host: 10.169.72.169:8080
Accept: */*


13:05:44.868329 IP 10.169.72.169.webcache > 10.169.72.93.57402: Flags [P.], seq 1:119, ack 82, win 4380, options [nop,nop,TS val 361318307 ecr 595756839], length 118: HTTP: HTTP/1.0 302 Found out slot1/tmm2 lis=/test/ingress_10-169-72-169_8080
E.....@....u
.H.
.H]...:......=............
..G.#..'HTTP/1.0 302 Found
Location: https://10.169.72.169:443/  <-------REDIRECT HERE
Server: BigIP
Connection: Keep-Alive
Content-Length: 0

</pre>

the LTM virtual and irule on BIGIP:

<pre>
ltm virtual ingress_10-169-72-169_443 {
    destination 10.169.72.169%0:https
    ip-protocol tcp
    mask 255.255.255.255
    metadata {
        user_agent {
            value k8s-bigip-ctlr-1.7.1-n1279-465125010
        }
    }
    partition test
    pool ingress_default_nginxservice
    profiles {
        /Common/clientssl {
            context clientside
        }
        /Common/http { }
        /Common/tcp { }
    }
    source 0.0.0.0/0
    source-address-translation {
        type automap
    }
    translate-address enabled
    translate-port enabled
    vs-index 34
}
ltm virtual ingress_10-169-72-169_8080 {
    destination 10.169.72.169%0:webcache
    ip-protocol tcp
    mask 255.255.255.255
    metadata {
        user_agent {
            value k8s-bigip-ctlr-1.7.1-n1279-465125010
        }
    }
    partition test
    pool ingress_default_nginxservice
    profiles {
        /Common/http { }
        /Common/tcp { }
    }
    rules {
        http_redirect_irule_443
    }
    source 0.0.0.0/0
    source-address-translation {
        type automap
    }
    translate-address enabled
    translate-port enabled
    vs-index 33
}

ltm rule http_redirect_irule_443 {
    metadata {
        user_agent {
            app-service none
            persist true
            value k8s-bigip-ctlr-1.7.1-n1279-465125010
        }
    }
    partition test
when HTTP_REQUEST {
			# Look for exact match for host name
			set paths [class match -value [HTTP::host] equals https_redirect_dg]
			if {$paths == ""} {
				# See if there's an entry that matches all hosts
				set paths [class match -value "*" equals https_redirect_dg]
			}
			if {$paths != ""} {
				set redir 0
				foreach s [split $paths "|"] {
					# See if the request path starts with the prefix
					append prefix "^" $s "($|/*)"
					if {[HTTP::path] matches_regex $prefix} {
						set redir 1
						break
					}
				}
				if {$redir == 1} {
					HTTP::redirect https://[getfield [HTTP::host] ":" 1]:443[HTTP::uri]
				}
			}
		}
}
</pre>

**Comment by bogiva** [comment]
What does your https_redirect_dg data group contain?
Can you log the contents of ```[HTTP::host]```? I did it by changing the code of the irule via big-ip web ui and testing quickly before it gets overwritten by the controller :).

If I remember correctly it was ```the_correct_hostname:8080``` in my case and the line
```
set paths [class match -value [HTTP::host] equals https_redirect_dg]
```
was calculating ```paths``` to be empty.

**Comment by vincentmli** [comment]
Here is my data-group
<pre>

ltm data-group internal https_redirect_dg {
    partition test
    records {
        \\\* {
            data /
        }
    }
    type string
}
</pre>

the paths will be empty
<pre>
			set paths [class match -value [HTTP::host] equals https_redirect_dg]
</pre>

but following should set the paths  since there is the "*" match all host entry	

<pre>

		if {$paths == ""} {
				# See if there's an entry that matches all hosts
				set paths [class match -value "*" equals https_redirect_dg]
			}
</pre>

could you please paste your ingress yaml file here so I can reproduce the issue?

**Comment by bogiva** [comment]
Exactly - the asterisk entry in your data group is the answer why it works at your side and why it doesn't at mine :).
I do not have it in my group. How did you come up with such a data group content?

Mine looks like this:
```
ltm data-group internal /partition1/https_redirect_dg {
    records {
        site-1.internal.domain1.com {
            data /
        }
        site-2.internal.domain1.com {
            data /
        }
        another-site.internal.domain1.com {
            data /
        }
    }
    type string
}
```

Here is how one of the ingresses look like:
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: ingress-write-api
  namespace: default
  annotations:
    virtual-server.f5.com/ip: "controller-default"
    virtual-server.f5.com/http-port: "8080"
    virtual-server.f5.com/partition: "partition1"
    virtual-server.f5.com/balance: "round-robin"
    virtual-server.f5.com/health: |
      [
        {
          "path":     "site-1.internal.domain1.com/",
          "send":     "GET / HTTP/1.1\r\nHost: site-1.internal.domain1.com\r\nConnection: Close",
          "interval": 10,
          "timeout":  20
        }
      ]
    ingress.kubernetes.io/ssl-redirect: "true"
    ingress.kubernetes.io/allow-http: "false"
    kubernetes.io/ingress.class: "f5"
spec:
  tls:
  - hosts:
    - site-1.internal.domain1.com
    secretName: site-1-internal-domain1-com-tls-secret
  rules:
  - host: site-1.internal.domain1.com
    http:
      paths:
      - path: /
        backend:
          serviceName: some-server
          servicePort: 8080
```

host names were anonymized.

**Comment by vincentmli** [comment]
I modified my ingress yaml like yours, I noticed my previous yaml does not have "rules:" like yours, here is my yaml, notice that my k8s nginxservice listening on port 80, not 8080, I did specify http-port annotation, I don't think the actual k8s nginxservice listening port 80 matter here, just want to mention the difference.

<pre>
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: single-service
  annotations:
     ingress.kubernetes.io/ssl-redirect: "true"
     ingress.kubernetes.io/allow-http: "false"
     ingress.kubernetes.io/allow-https: "true"
     kubernetes.io/ingress.class: f5
     'virtual-server.f5.com/ip': '10.169.72.169'
     'virtual-server.f5.com/http-port': '8080'
spec:
  tls:
  - hosts:
    - www.example.com
    secretName: Common/clientssl
  rules:
  - host: www.example.com 
    http:
       paths:
       - path: /
         backend:
           serviceName: nginxservice 
           servicePort: 80
</pre>

here is my full ltm config on BIGIP

<pre>

ltm data-group internal https_redirect_dg {
    partition test
    records {
        www.example.com {
            data /
        }
    }
    type string
}

ltm policy ingress_10-169-72-169_443 {
    controls { forwarding }
    last-modified 2019-01-07:09:43:38
    partition test
    requires { http }
    rules {
        ingress_www.example.com__ingress_default_nginxservice {
            actions {
                0 {
                    forward
                    select
                    pool ingress_default_nginxservice
                }
            }
            conditions {
                0 {
                    http-host
                    host                    values { www.example.com }
                }
            }
        }
    }
    status legacy
    strategy /Common/first-match
}
ltm policy ingress_10-169-72-169_8080 { <==== this policy act before http_redirect_irule_443 
    controls { forwarding }
    last-modified 2019-01-07:09:43:38
    partition test
    requires { http }
    rules {
        ingress_www.example.com__ingress_default_nginxservice {
            actions {
                0 {
                    forward
                    select
                    pool ingress_default_nginxservice
                }
            }
            conditions {
                0 {
                    http-host
                    host
                    values { www.example.com }
                }
            }
        }
    }
    status legacy
    strategy /Common/first-match
}
ltm pool ingress_default_nginxservice {
    members {
        10.244.0.60%0:http {
            address 10.244.0.60
        }
        10.244.0.61%0:http {
            address 10.244.0.61
        }
    }
    metadata {
        user_agent {
            value k8s-bigip-ctlr-1.7.1-n1279-465125010
        }
    }
    partition test
}
ltm rule http_redirect_irule_443 {
    metadata {
        user_agent {
            app-service none
            persist true
            value k8s-bigip-ctlr-1.7.1-n1279-465125010
        }
    }
    partition test
when HTTP_REQUEST {
                        # Look for exact match for host name
                        set paths [class match -value [HTTP::host] equals https_redirect_dg]
                        if {$paths == ""} {
                                # See if there's an entry that matches all hosts
                                set paths [class match -value "*" equals https_redirect_dg]
                        }
                        if {$paths != ""} {
                                set redir 0
                                foreach s [split $paths "|"] {
                                        # See if the request path starts with the prefix
                                        append prefix "^" $s "($|/*)"
                                        if {[HTTP::path] matches_regex $prefix} {
                                                set redir 1
                                                break
                                        }
                                }
                                if {$redir == 1} {
                                        HTTP::redirect https://[getfield [HTTP::host] ":" 1]:443[HTTP::uri]
                                }
                        }
                }
}

ltm virtual ingress_10-169-72-169_443 {
    destination 10.169.72.169%0:https
    ip-protocol tcp
    mask 255.255.255.255
    metadata {
        user_agent {
            value k8s-bigip-ctlr-1.7.1-n1279-465125010
        }
    }
    partition test
    policies {
        ingress_10-169-72-169_443 { }
    }
    profiles {
        /Common/clientssl {
            context clientside
        }        /Common/http { }
        /Common/tcp { }
    }
    source 0.0.0.0/0
    source-address-translation {
        type automap
    }
    translate-address enabled
    translate-port enabled
    vs-index 42
}
ltm virtual ingress_10-169-72-169_8080 {
    destination 10.169.72.169%0:webcache
    ip-protocol tcp
    mask 255.255.255.255
    metadata {
        user_agent {
            value k8s-bigip-ctlr-1.7.1-n1279-465125010
        }
    }
    partition test
    policies {
        ingress_10-169-72-169_8080 { }
    }
    profiles {
        /Common/http { }
        /Common/tcp { }
    }
    rules {
        http_redirect_irule_443
    }
    source 0.0.0.0/0
    source-address-translation {
        type automap
    }
    translate-address enabled
    translate-port enabled
    vs-index 41
}
</pre>

if I run curl http://www.example.com:8080/, it works because ltm policy ingress_10-169-72-169_8080 kicks in before the redirect irule, I think this might be a different bug since I have ingress.kubernetes.io/ssl-redirect: "true" and ingress.kubernetes.io/allow-http: "false" in my yaml

<pre>
[root@centos-k8s-master kubernetes]# curl http://www.example.com:8080
<!DOCTYPE html>
<html>

<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>

</html>
</pre>

so I can't reproduce your issue, I am curious to know if you have any ltm policy created? could you share your full relevant ltm configuration on BIGIP when you deployed your yaml?

**Comment by vincentmli** [comment]
ok glad the problem is clear now, I think the fix is not to set policy on the http virtual server, just a http to https redirect irule when ssl-redirect: true and allow-http: false. I read the code and attempted to work on a fix, it is bit of messy, there are datagroup for irule, policy setup based on ingress rules...etc. 

the policy setup is  for ingress with "rules:" defined, this explain why I have no policy setup with ingress with no "rules:" defined. as a workaround, could you adjust your yaml without "rules:" as below? this way, the ingress is treated as single service and there will be no policy setup, and you will only have the wildcard datagroup defined for the redirect irule, let me know if that could meet your requirement as temporary solution.

<pre>

apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: some-ingress
  namespace: default
  annotations:
    virtual-server.f5.com/ip: "controller-default"
    virtual-server.f5.com/http-port: "8080"
    virtual-server.f5.com/partition: "partition1"
    virtual-server.f5.com/balance: "round-robin"
    virtual-server.f5.com/health: |
      [
        {
          "path":     "site-1.internal.domain1.com/",
          "send":     "GET / HTTP/1.1\r\nHost: site-1.internal.domain1.com\r\nConnection: Close",
          "interval": 10,
          "timeout":  20
        }
      ]
    ingress.kubernetes.io/ssl-redirect: "true"
    ingress.kubernetes.io/allow-http: "false"
    kubernetes.io/ingress.class: "f5"
spec:
  tls:
  - hosts:
    - site-1.internal.domain1.com
    secretName: site-1-internal-domain1-com-tls-secret
  backend:
        serviceName: some-server
        servicePort: 8080
</pre>

**Comment by vincentmli** [comment]
here is a patch to fix http to https redirect issue when following three condition meet:
 <pre>
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: single-service
  annotations:
     ingress.kubernetes.io/ssl-redirect: "true" < 1----SSL redirect true
     ingress.kubernetes.io/allow-http: "false" <  2----allow-http false
     ingress.kubernetes.io/allow-https: "true"
     kubernetes.io/ingress.class: f5
     'virtual-server.f5.com/ip': '10.169.72.169'
spec:
  tls:
  - hosts:
    - www.example.com
    secretName: Common/clientssl
  rules: < 3 --------ingress with rules defined 
  - host: www.example.com 
    http:
       paths:
       - path: /
         backend:
           serviceName: nginxservice 
           servicePort: 80
  
</pre>

the patch is not to set policy, but set http to https redirect irule to the http virtual server
<pre>
ltm virtual ingress_10-169-72-169_80 {
    destination 10.169.72.169%0:http
    ip-protocol tcp
    mask 255.255.255.255
    metadata {
        user_agent {
            value k8s-bigip-ctlr-1.8.0-WIP-vli-d029d23-dirty-20190110095900
        }
    }
    partition test
    profiles {
        /Common/http { }
        /Common/tcp { }
    }
    rules {
        http_redirect_irule_443
    }
    source 0.0.0.0/0
    source-address-translation {
        type automap
    }
    translate-address enabled
    translate-port enabled
    vs-index 42
}

</pre>

has to copy the SetPolicy function to a new IngSetPolicy and insert the sslRedirect condition there, changing SetPolicy directly result in messy healthMonitors_test.go  test problem ( why would SetPolicy affect health monitor test? probably the health unit test code need to be re-evaluated. )

<pre>

diff --git a/pkg/appmanager/resourceConfig.go b/pkg/appmanager/resourceConfig.go
index e7b06fa..b764c51 100644
--- a/pkg/appmanager/resourceConfig.go
+++ b/pkg/appmanager/resourceConfig.go
@@ -1573,6 +1573,9 @@ func (appMgr *Manager) createRSConfigFromIngress(
        }
        cfg.MetaData.ingName = ing.ObjectMeta.Name
 
+       sslRedirect := getBooleanAnnotation(ing.ObjectMeta.Annotations,
+               ingressSslRedirect, true)
+
        resources.Lock()
        defer resources.Unlock()
        // Check to see if we already have any Ingresses for this IP:Port
@@ -1619,7 +1622,7 @@ func (appMgr *Manager) createRSConfigFromIngress(
                                }
                        }
                } else if len(cfg.Policies) == 0 && plcy != nil {
-                       cfg.SetPolicy(*plcy)
+                       cfg.IngSetPolicy(*plcy, sslRedirect, pStruct.protocol)
                }
        } else { // This is a new VS for an Ingress
                cfg.MetaData.ResourceType = "ingress"
@@ -1629,7 +1632,7 @@ func (appMgr *Manager) createRSConfigFromIngress(
                cfg.Virtual.SetVirtualAddress(bindAddr, pStruct.port)
                cfg.Pools = append(cfg.Pools, pools...)
                if plcy != nil {
-                       cfg.SetPolicy(*plcy)
+                       cfg.IngSetPolicy(*plcy, sslRedirect, pStruct.protocol)
                }
        }
 
@@ -2209,6 +2212,36 @@ func (rc *ResourceConfig) SetPolicy(policy Policy) {
        rc.Policies = append(rc.Policies, policy)
 }
 
+func (rc *ResourceConfig) IngSetPolicy(policy Policy, sslRedirect bool, pProtocol string) {
+       toFind := nameRef{
+               Name:      policy.Name,
+               Partition: policy.Partition,
+       }
+       found := false
+       for _, polName := range rc.Virtual.Policies {
+               if reflect.DeepEqual(toFind, polName) {
+                       found = true
+                       break
+               }
+       }
+       if !found {
+               if sslRedirect {
+                       if pProtocol == "https" {
+                               rc.Virtual.Policies = append(rc.Virtual.Policies, toFind)
+                       }
+               } else {
+                       rc.Virtual.Policies = append(rc.Virtual.Policies, toFind)
+               }
+       }
+       for i, pol := range rc.Policies {
+               if pol.Name == policy.Name && pol.Partition == policy.Partition {
+                       rc.Policies[i] = policy
+                       return
+               }
+       }
+       rc.Policies = append(rc.Policies, policy)
+}
+
</pre>

I also  tested sslRedirect to be false, allow-http true, both http and https has proper policy setup and traffic test passing ok.

**Comment by vincentmli** [comment]
the second problem of this issue is when http-port is set to non port 80, the http to https redirect would also fail, for example:

<pre>
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: single-service
  annotations:
     ingress.kubernetes.io/ssl-redirect: "true"
     ingress.kubernetes.io/allow-http: "false"
     ingress.kubernetes.io/allow-https: "true"
     kubernetes.io/ingress.class: f5
     'virtual-server.f5.com/ip': '10.169.72.169'
     'virtual-server.f5.com/http-port': '8080' <----set http-port to 8080
spec:
  tls:
  - hosts:
    - www.example.com
    secretName: Common/clientssl
  rules:
   - host: www.example.com
     http:
       paths:
       - path: /
         backend:
            serviceName: nginxservice
            servicePort: 80
</pre>

it turned out there is just one-line change fix to http_redirect_irule_443  to fix the second problem, the datagroup https_redirect_dg  contains hostname, so when [HTTP::host] contains "hostname:port", the irule should only take the hostname, not the port, thus [getfield [HTTP::host] ":" 1]  to get the hostname

<pre>
diff --git a/pkg/appmanager/routing.go b/pkg/appmanager/routing.go
index 1a450c4..6679403 100644
--- a/pkg/appmanager/routing.go
+++ b/pkg/appmanager/routing.go
@@ -350,7 +350,7 @@ func httpRedirectIRule(port int32) string {
        iRuleCode := fmt.Sprintf(`
                when HTTP_REQUEST {
                        # Look for exact match for host name
-                       set paths [class match -value [HTTP::host] equals https_redirect_dg]
+                      set paths [class match -value [getfield [HTTP::host] ":" 1] equals https_redirect_dg]
                        if {$paths == ""} {
                                # See if there's an entry that matches all hosts
                                set paths [class match -value "*" equals https_redirect_dg]
</pre>

test:

<pre>
# curl -k -L http://www.example.com:8080/
Welcome to nginx!
</pre>

**Comment by bogiva** [comment]
@vincentmli, the suggested fix that takes the hostname (without the port) seems like the right thing to do. Which release will have the fix :)?

**Comment by vincentmli** [comment]
to fully address the proper http to https redirect issue, both fixes I posted here need to addressed, 1 remove policy configuration for http virtual server, 2 fix the hostname:port in http_redirect_irule_443, I can discuss with the release team about this matter.

**Comment by mdditt2000** [comment]
serviceMain-Redirect is enabled by default now in AS3. I validated this. CIS 1.9 released next month.

**Comment by mdditt2000** [comment]
Resolved in CIS 1.9 with AS3. 

![image](https://user-images.githubusercontent.com/12733938/55189387-050b2100-515b-11e9-9584-667bc7a2481d.png)

**Comment by vincentmli** [comment]
this issue still not resolved by CIS + AS3 when the http port is not standard port 80. as my previous notes, there is really no need to have policy configured for http virtual server when ssl redirect is true
